DOCS-5467-Access-Management-Connected-Apps

sarathecoubian opened the pull request:
Information can be moved wherever needed. Suggestions encouraged. Please note in the text files that there are comments denoting where screenshots and blocks of code should go.

sarathecoubian committed:
DOCS-5467-Access-Management-Connected-Apps

Information can be moved wherever needed. Suggestions encouraged.

sarathecoubian committed:
Update connected-apps-overview.adoc

valkyrie69 commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+
+== Authentication
+
+Using OAuth 2.0 and OpenID, you can authorize applications to access resources using tokens or Single Sign-On.
```
```suggestion
Use OAuth 2.0 and OpenID to authorize apps to access resources using tokens or Single Sign-On (SSO).
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+
+Using OAuth 2.0 and OpenID, you can authorize applications to access resources using tokens or Single Sign-On.
+Additionally, you can share account data with third-party applications.
+For more information on which users can grant access to applications using Anypoint Platform, see Connected Apps Permissions by User.
```
Are you adding a link to that topic in See Also?

sarathecoubian replied:
I can.

valkyrie69 commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+include::_attributes.adoc[]
+endif::[]
+
+You can connect applications to Anypoint Platform without granting them unrestricted access to data or having to share sensitive credentials.
```
I think our style guide says to use app instead of application.

sarathecoubian replied:
Change made throughout.

valkyrie69 commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+== First-party and Third-party Applications
+
+Depending on the ownership of the application you want to connect, applications may have different authentication requirements and procedures.
+To determine whether the application you want to connect is a first-party or third party application, consider the following:
```
```suggestion
To determine whether the app you want to connect is a first- or third-party application, consider the below:
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+Depending on the ownership of the application you want to connect, applications may have different authentication requirements and procedures.
+To determine whether the application you want to connect is a first-party or third party application, consider the following:
+
+* First-party: an application that is created and controlled by the same entity as the domain owner.
```
```suggestion
* First-party
   An application that is created and controlled by the same entity as the domain owner.
```

sarathecoubian replied:
```suggestion
* First-party:: 
an app that is controlled by organization to which you belong.
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+To determine whether the application you want to connect is a first-party or third party application, consider the following:
+
+* First-party: an application that is created and controlled by the same entity as the domain owner.
+* Third-party application: an application that is controlled by a different entity than the domain owner.
```
```suggestion
* Third-party app
   An application that is controlled by a different entity than the domain owner.
```

sarathecoubian replied:
```suggestion
* Third-party application:: 
An app that is controlled by another organization.

Applications can also be public or private.

* Public applications can be accessed by anyone.
* Private applications can only be accessed by users in the same organization.
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+include::_attributes.adoc[]
+endif::[]
+
+Connected Apps can be used by Organization Administrators, developers, and end-users; each of these roles has different privileges when connecting applications,
```
```suggestion
Connected apps can be used by organization administrators, developers, and end users; each of these roles has different privileges when connecting applications,
```
I'm not sure about organization administrators. That is an actual role so it might need to remain title case and be bolded.  Kim will know.

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+== Organization Administrator
+
+An Organization Administrator is the only role that can view and manage Connected Apps in Access Management.
```
I'm not sure about capping Connected Apps. I would lc all, but Kim might have diff advice.

sarathecoubian replied:
```suggestion
An Organization Administrator is the only role that can view and manage connected apps in Access Management.
```

sarathecoubian replied:
I don't think it should be capped unless I'm referring to the page at this point.

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+// screenshot of Connected Apps page
+
+The *Connected Apps* page shows a list of applications that you have connected to Anypoint Platform.
```
```suggestion
The *Connected Apps* page shows a list of apps that are connected to Anypoint Platform.
```

valkyrie69 replied:
Technically, this is a procedure and should be an ol.

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+The *Connected Apps* page shows a list of applications that you have connected to Anypoint Platform.
+To add an application, select *Add Application*.
+
+=== Create Application
```
Create an App

sarathecoubian replied:
```suggestion
=== Create an App
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+=== Create Application
+
+When you create an application as an Organization Administrator, consider the following:
```
```suggestion
When you create an app as an Organization Administrator, consider the below:
```

sarathecoubian replied:
```suggestion
When you create an app as an organization administrator, consider the below:
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+* An organization can own up to 200 connected apps.
+* You can disable or whitelist connected apps.
+* An application is assigned a clientID upon creation; the clientID cannot be modified. However, client secrets can be modified in the *Application Settings* page.
```
```suggestion
* When you create an app, a clientID is assigned, which can't be modified. However, you can modify client secrets in the *Application Settings* page.
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+* An organization can own up to 200 connected apps.
+* You can disable or whitelist connected apps.
+* An application is assigned a clientID upon creation; the clientID cannot be modified. However, client secrets can be modified in the *Application Settings* page.
+* When you create an application with the client grant type, the application will continue to work if the application owner's credentials are disabled for any reason. A team or company can assume control of the application's access.
```
```suggestion
* When you create an application with the OAuth client credential grant type, the application continues to work even if the application owner's credentials are disabled for any reason. A team or company can assume control of the application's access.
```

maxwellgerber replied:
There is no such thing as an application owner for Connected Apps. ConnApps are owned by an organization, and any Org Admin in the administration can control the app.

sarathecoubian replied:
```suggestion
* When you create an application with the OAuth client credential grant type, the app is owned by the organization; any organization administrator can control the app.
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+* You can disable or whitelist connected apps.
+* An application is assigned a clientID upon creation; the clientID cannot be modified. However, client secrets can be modified in the *Application Settings* page.
+* When you create an application with the client grant type, the application will continue to work if the application owner's credentials are disabled for any reason. A team or company can assume control of the application's access.
+* An application's actions are logged with an application ID and user ID (for user grant type applications only) in the Audit Log. This empowers administrators to monitor and audit data use.
```
Please verify the grant type. AFAIK, there is not a "user" grant type for OAuth. It would be a **password** grant type, where the user name and password are passed.

maxwellgerber replied:
We've been using the terminology User Grant Type to represent the grant types that require user interaction ('implicit', 'authorization_code', 'refresh_token') and Client Grant Type to represent the grant types that do not ('client_credentials').

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+// screenshot of Create Application interface
+
+When creating an application, you must enter the following information:
```
```suggestion
When creating an application, you must enter the below information:
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+* You can disable or whitelist connected apps.
+* An application is assigned a clientID upon creation; the clientID cannot be modified. However, client secrets can be modified in the *Application Settings* page.
+* When you create an application with the client grant type, the application will continue to work if the application owner's credentials are disabled for any reason. A team or company can assume control of the application's access.
+* An application's actions are logged with an application ID and user ID (for user grant type applications only) in the Audit Log. This empowers administrators to monitor and audit data use.
```
```suggestion
* For apps that use the password grant type, actions taken by the app are logged with an application ID and user ID in the audit log so administrators can monitor and audit data use.
```

maxwellgerber replied:
I'd avoid mentioning password grant type - that's a separate thing from client_credentials that we haven't implemented for Connected Apps (And don't plan on doing so either!)

sarathecoubian replied:
```suggestion
* For apps that use the user grant type, actions taken by the app are logged with an application ID and user ID in the audit log so administrators can monitor and audit data use.
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+When creating an application, you must enter the following information:
+
+* *Application Name*: A descriptive name for your application.
```
Kim wants us to do list items like this:

- Item
   Definition
- Item 
   Definition

Basically, like a DL

sarathecoubian replied:
```suggestion
* *Application Name*:: 
A descriptive name for your application.
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+* *Application Name*: A descriptive name for your application.
+* *Redirect URIs*: Uniform Resource Identifier that specifies the location of the application. You can specify one per line.
+* *User Grant Types* or *Client Grant Types*: A grant type that aligns with OAuth 2.0 specifications.
```
This looks like UI, so there must be a user grant type. But please do verify what that means because, again, in the world of OAuth 2.0 I don't know of a "user" grant type. Not that I'm an expert.

valkyrie69 replied:
Just verify that if it is indeed a "user" grant type, does that mean the user name and password are passed, or...?

maxwellgerber replied:
Mentioned above - we are not supporting the password grant type. We've been using user grant types as a term to mean Grant Types that require user input

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+==== User Grants
+
+User grants require authentication from a user, and they are primarily useful for third-party applications. Connected Apps currently supports two types of user grant authentication:
```
Authentication meaning user name and password?

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+User grants require authentication from a user, and they are primarily useful for third-party applications. Connected Apps currently supports two types of user grant authentication:
+
+* Authorization: requires the application to produce an authorization code that is accepted by OAuth 2.0.
```
This is bugging me...I guess they are grouping diff grant types under "user." Strange. Because authorization code is an actual grant type.  I'm not commenting on this anymore.  :-)

maxwellgerber replied:
```suggestion
* Browser-based Authorization Flow::
 The app directs the user's browser (commonly called a User-Agent) to Anypoint, where the user authorizes the app before being redirected back
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+User grants require authentication from a user, and they are primarily useful for third-party applications. Connected Apps currently supports two types of user grant authentication:
+
+* Authorization: requires the application to produce an authorization code that is accepted by OAuth 2.0.
+If the code should not work indefinitely, you can select the *Refresh Tokens* option, requiring new codes as often as specified in the application.
```
```suggestion
Select the *Refresh Tokens* option to require new codes as often as specified in the application rather than allowing code to work indefinitely.
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+* Authorization: requires the application to produce an authorization code that is accepted by OAuth 2.0.
+If the code should not work indefinitely, you can select the *Refresh Tokens* option, requiring new codes as often as specified in the application.
+* JTW Bearer: grants access using an Application Public Key in a Privacy Enhanced Mail (PEM) that validates the JWT Authorization.
```
```suggestion
* JTW Bearer
  Grants access using an application public key in a Privacy-Enhanced Mail (PEM) that validates 
  the JWT Authorization.
```

maxwellgerber replied:
JWT Bearer apps are private apps that the Organization Administrator registers with a public key. The App then uses the corresponding private key to sign JWTs that may then be exchanged for an access token. When an Organization Administrator registers the app, they authorize it to retrieve tokens for all users in the org. Because of this, it is still considered a user grant type, even though no users will go through the authorization flow. In addition, individual users cannot revoke access to their account.
I wouldn't mention PEM files - we just have private and public keys.

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+If the code should not work indefinitely, you can select the *Refresh Tokens* option, requiring new codes as often as specified in the application.
+* JTW Bearer: grants access using an Application Public Key in a Privacy Enhanced Mail (PEM) that validates the JWT Authorization.
+
+After you select one or more user grant types, you can specify the scope of permissions.
```
```suggestion
You can specify the scope of permissions after you select one or more user grant types.
```

maxwellgerber replied:
Scope of permissions is a confusing turn of phrase.
For apps using a user grant type, you specify the list of scopes the app is allowed to request from a user. Each scope represents a set of permissions, identical to how a role represents a set of permissions. Unlike roles, you do not assign the scopes with context, since the context comes from the user who authorizes the app.

For apps using a client grant type, you assign scopes with a context. These assignments can be thought of as identical to assigning an app a role directly. (And they are! just in a OAuth 2.0 flavored / spec compliant way)

sarathecoubian replied:
```suggestion
You can specify the list of scopes the app is allowed to request from a user after you select one or more user grant types.
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+==== Client Grants
+
+Client grants are only permitted for first-party applications; they give users access based on their organizational affiliations.
```
```suggestion
Client grants give users access based on their organizational affiliations and are permitted only for first-party apps.
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+== Developer
+
+Developers who have websites or web apps can allow users to log in using the *Log in with Anypoint Platform* option.
```
I'm not sure web sites is one word.

valkyrie69 replied:
Also, this is a UI thing, but they should change that UI label to Sign in instead of Log in. The nomenclature on the Anypoint Platform sign-in page is sign in.

sarathecoubian replied:
I shall bring it up. From what I can tell, websites is a single word, but I imagine Kim will have an opinion.

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+== End-user
+
+An end-user can log into a connected application using their Single Sign-On credentials. End-users can authorize access when the following message appears:
```
I don't know if end-user is hyphenated. I think it's just end user, but Kim will know for sure.

sarathecoubian replied:
```suggestion
An end user can log into a connected application using their Anypoint credentials. End users can also  authorize access when the following message appears:
```

valkyrie69 commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+// screenshot of authorize user page https://qax.anypoint.mulesoft.com/login/authorize?state=eyJjbGllbnQiOnsiY2xpZW50X2lkIjoiZWEzYzJiODFkMDEyNGI0Mjk1NDYyM2RhMWM0MzQ5ZWIifSwidXNlciI6eyJ1c2VybmFtZSI6Im9yZ2FuaXphdGlvbl9vd25lciJ9LCJvcmlnaW4iOiIvYXBpL3YyL29hdXRoMi9hdXRob3JpemU%2FY2xpZW50X2lkPWVhM2MyYjgxZDAxMjRiNDI5NTQ2MjNkYTFjNDM0OWViJnJlZGlyZWN0X3VyaT1odHRwczovL2V4YW1wbGUuY29tJnJlc3BvbnNlX3R5cGU9dG9r ...
+
+End-users can manage third-party applications to which they have granted access in their profile settings.
```
```suggestion
End-users can manage third-party applications to which they have access in their profile settings.
```

Should it be **Profile** settings? Is it UI?

maxwellgerber commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+
+== First-party and Third-party Applications
+
+Depending on the ownership of the application you want to connect, applications may have different authentication requirements and procedures.
```
> may have different authentication requirements and procedures

Authentication requirements are always the same (The user must be logged in).
Procedures - We originally wanted to support auto-approval for first-party apps but that was culled as a feature a few months back. There aren't any other procedural differences I'm aware of

sarathecoubian replied:
```suggestion
```

maxwellgerber commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+Depending on the ownership of the application you want to connect, applications may have different authentication requirements and procedures.
+To determine whether the application you want to connect is a first-party or third party application, consider the following:
+
+* First-party: an application that is created and controlled by the same entity as the domain owner.
```
`same entity as the domain owner` reads funny to me
- First-party applications are controlled by the organization you belong to
- Third-party applications are controlled by another organization 

Applications can also be public or private
- Public applications can be accessed by anyone
- Private applications can only be accessed by users in the same organization

So all third-party applications are necessarily public applications
and all private applications are necessarily first-party applications
but depending on the user, a public application might be first party or third party

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+== Organization Administrator
+
+An Organization Administrator is the only role that can view and manage Connected Apps in Access Management.
```
```suggestion
Organization administrators are the only users that can view and manage connected apps in Access Management.
```

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+When you create an application as an Organization Administrator, consider the following:
+
+* An organization can own up to 200 connected apps.
+* You can disable or whitelist connected apps.
```
Whitelisting connected apps is orthogonal to their creation.

We have two personas:
**A developer** that wants to register and administrate an Application that they've made. 
This includes:
- setting grant types, client names, redirect uris
- Enabling/Disabling/Deleting the client if its secret has been leaked
- Viewing usage about their client (How many organizations are using it? How many users?)

**An Administrator** that wants to control how their organization's data is exposed.
This includes:
- Viewing which users have shared their data to which apps, and revoking share access if necessary
- Maintaining the list of apps that users are allowed to share data to (The whitelist)
- Maintaining settings about how data can be shared
    - What grant types can be used? Is the whitelist enabled?

Today, both of those personas are performed by Organization Administrators. But, I'd be very deliberate about separating the two when talking about them.

A more concrete example: Github
I own one Github app. To see it, I navigate to Developer Settings
![image](https://user-images.githubusercontent.com/11035210/63310888-3b377180-c2b1-11e9-8387-06fc6a1c27e0.png)

I've also authorized that Github app to access my data. To see that, I navigate to my Personal Settings
![image](https://user-images.githubusercontent.com/11035210/63310931-5c985d80-c2b1-11e9-9f19-7512f0612c0e.png)

I am both a producer of my app and a consumer. But, my app is still treated like everyone else's. The majority of our users will only be app consumers, not creators.

sarathecoubian replied:
I'm going to follow up with you on explaining this a little more carefully to users - I am still pretty unsure of where devs fit into this equation.

sebastiankorol replied:
@sarathecoubian this is the most important piece of feedback. There's org admins creating and managing applications on one side, and there's users connecting applications, and their org admins managing their authorizations and whitelists on the other. The two are very distinct and I believe they need separate pages.

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+When creating an application, you must enter the following information:
+
+* *Application Name*: A descriptive name for your application.
+* *Redirect URIs*: Uniform Resource Identifier that specifies the location of the application. You can specify one per line.
```
```suggestion
* *Redirect URIs*::
 Uniform Resource Identifier that specifies the location of the application's callback. You can specify one per line.
```

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+* *Application Name*: A descriptive name for your application.
+* *Redirect URIs*: Uniform Resource Identifier that specifies the location of the application. You can specify one per line.
+* *User Grant Types* or *Client Grant Types*: A grant type that aligns with OAuth 2.0 specifications.
```
```suggestion
* *Allowed Grant Types*: A list of OAuth 2.0 grant types the application intends to make use of
```

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+* *Application Name*: A descriptive name for your application.
+* *Redirect URIs*: Uniform Resource Identifier that specifies the location of the application. You can specify one per line.
+* *User Grant Types* or *Client Grant Types*: A grant type that aligns with OAuth 2.0 specifications.
+It is important to specify the correct grant type when connecting an application since it will dictate who can access the application. Note that client grant types are only available for first-party applications.
```
It's the other way around - who can access the application dictates what grant types are allowed. If an application is made public, client grant types are not allowed.
First party applications, if made public, can't have the client grant type - since if they are made public they are someone else's third party application

sarathecoubian replied:
```suggestion
It is important to specify the access permissions when connecting an app since that will dictate the available grant types. Note that client grant types are only available for private first-party apps. 
```

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+==== User Grants
+
+User grants require authentication from a user, and they are primarily useful for third-party applications. Connected Apps currently supports two types of user grant authentication:
```
```suggestion
Clients making use of User grant types require authorization from a user. Connected Apps currently supports two methods of receiving authorization.
```

sarathecoubian replied:
```suggestion
Clients making use of user grant types require authorization from a user. Connected Apps currently supports two methods of receiving authorization.
```

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+User grants require authentication from a user, and they are primarily useful for third-party applications. Connected Apps currently supports two types of user grant authentication:
+
+* Authorization: requires the application to produce an authorization code that is accepted by OAuth 2.0.
+If the code should not work indefinitely, you can select the *Refresh Tokens* option, requiring new codes as often as specified in the application.
```
- Codes must be exchanged for access tokens
- they can only be exchanged once
- access tokens ALWAYS expire within a short period of time, usually a matter of hours, or when the user logs out

If the app requires access after the user logs out, or doesn't want to prompt them every time, they can use refresh tokens.
- The app must have the authorization_code grant type to use the refresh_token grant type
- Codes _may_ be exchanged for an (access token, refresh token) pair iff the user approves it
- refresh tokens may be exchanged by the app at a later date for another (access token, refresh token) pair

sarathecoubian replied:
How much does a user already know about refresh tokens?

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+==== Client Grants
+
+Client grants are only permitted for first-party applications; they give users access based on their organizational affiliations.
+You can also create more granular access restrictions by specifying which environments have access.
```
Another weird turn of phrase - you specify which environments the app may access - not which environments may access the app

sarathecoubian replied:
```suggestion
You can also specify which environments the app can access.
```

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+== Developer
+
+Developers who have websites or web apps can allow users to log in using the *Log in with Anypoint Platform* option.
+For this feature to work, the connected application must be configured with user grant authentication.
```
We're still fleshing out the details on what Login with Anypoint entails, but there's a few more things coming 
The one I can guarantee is that the app must be configured with the `openid` scope at bare minimum

sarathecoubian replied:
I will follow up with you on this.

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+An end-user can log into a connected application using their Single Sign-On credentials. End-users can authorize access when the following message appears:
+
+// screenshot of authorize user page https://qax.anypoint.mulesoft.com/login/authorize?state=eyJjbGllbnQiOnsiY2xpZW50X2lkIjoiZWEzYzJiODFkMDEyNGI0Mjk1NDYyM2RhMWM0MzQ5ZWIifSwidXNlciI6eyJ1c2VybmFtZSI6Im9yZ2FuaXphdGlvbl9vd25lciJ9LCJvcmlnaW4iOiIvYXBpL3YyL29hdXRoMi9hdXRob3JpemU%2FY2xpZW50X2lkPWVhM2MyYjgxZDAxMjRiNDI5NTQ2MjNkYTFjNDM0OWViJnJlZGlyZWN0X3VyaT1odHRwczovL2V4YW1wbGUuY29tJnJlc3BvbnNlX3R5cGU9dG9r ...
```
The Authorize page is changing soon @LeoWW let's make sure we keep the docs in sync with UI changes

0xTenable replied:
Ack -- I'll update Sara when UI features are pushed

0xTenable replied:
FYI: Here's the mockup for what the final Connected Apps UI will look like: https://www.figma.com/file/nGfK0OXqbSk15WoLN4dO2Q/Connected-apps?node-id=0%3A1

maxwellgerber commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+== End-user
+
+An end-user can log into a connected application using their Single Sign-On credentials. End-users can authorize access when the following message appears:
```
Another weird phrase for me - end users authenticate themselves to Anypoint using whatever means they have configured, whether that's SSO or regular username/password. Once authenticated to Anypoint they can authorize any apps they want.

maxwellgerber replied:
Connected Applications also allows Anypoint to be the SSO provider for these apps. So there's 1-2 levels of SSO, ex.
`Okta -> Anypoint -> My App` when external identity is configured for the user's org
`Anypoint -> My App` when the user is logging in with a username and password 
I'm not sure how to capture intent here

sarathecoubian replied:
Not really sure how or where to capture the second half of your comment; I am reorganizing this document's information, more than likely, so I will follow up with you to figure out where this should go.

maxwellgerber reviewed (commented):
Awesome!!
Left some initial feedback.
My biggest concern is clear communication about internal / external apps vs first/third party apps

Internal / External  or Public/Private is a property of the app. A Private app can only be accessed by users in the same organization. A Public app can be accessed by anyone.

First / Third Party is a property of an (app, user) pair. If the app and user are in the same org, it is a first party app. If the app and user are in different orgs, it is a third party app.

All private apps only have first party relations.
Public apps may have first OR third party relations.

In general, we ignore first / third party and treat all apps the same. 
Except for one special case - we have a setting to add internal apps to the whitelist automatically, but we aren't sure if/how we want to show that in the UI yet.

Please feel free to reach out with any more questions.

sarathecoubian commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+// code/chunk of steps for websites/web apps
+
+== End-user
```
```suggestion
== End User
```

sarathecoubian commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+include::_attributes.adoc[]
+endif::[]
+
+You can connect applications to Anypoint Platform without granting them unrestricted access to data or having to share sensitive credentials.
```
```suggestion
You can connect apps to Anypoint Platform without granting them unrestricted access to data or having to share sensitive credentials.
```

sarathecoubian commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+endif::[]
+
+You can connect applications to Anypoint Platform without granting them unrestricted access to data or having to share sensitive credentials.
+When you grant an application access, you have direct control over its data access, usage, and credentials.
```
```suggestion
When you grant an app access, you have direct control over its data access, usage, and credentials.
```

sarathecoubian commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+
+* Authorization: requires the application to produce an authorization code that is accepted by OAuth 2.0.
+If the code should not work indefinitely, you can select the *Refresh Tokens* option, requiring new codes as often as specified in the application.
+* JTW Bearer: grants access using an Application Public Key in a Privacy Enhanced Mail (PEM) that validates the JWT Authorization.
```
```suggestion
* JTW Bearer:: 
A private app that the organization administrator registers with a public key. The app uses the corresponding private key to sign JWTs that may then be exchanged for an access token. 
```

sarathecoubian commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,69 @@
+Client grants are only permitted for first-party applications; they give users access based on their organizational affiliations.
+You can also create more granular access restrictions by specifying which environments have access.
+
+After you select an organization or organization and environment, you can specify the scope of permissions.
```
```suggestion
You can specify the list of scopes the app is allowed to request from a user after you select an organization or organization and environment.
```

sarathecoubian committed:
Apply suggestions from code review

Need to follow up on some feedback - initial changes implemented as best as I know how.

Co-Authored-By: valkyrie69 <36547707+valkyrie69@users.noreply.github.com>
Co-Authored-By: Maxwell Gerber <maxwell.gerber@mulesoft.com>

sarathecoubian committed:
Formatting.

sarathecoubian committed:
Formatting.

0xTenable commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+
+* First-party: 
+
+  An app that is controlled by organization to which you belong.
```
nit:  An app that is controlled by an organization you belong to

sarathecoubian replied:
```suggestion
  An app that is controlled by your organization.
```

0xTenable commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,80 @@
+include::_attributes.adoc[]
+endif::[]
+
+Connected Apps can be used by organization administrators, developers, and end users; each of these roles has different privileges when connecting applications,
```
Nit: This feels like a run on sentence. I suggest using a period :) 

```
Connected Apps can be used by organization administrators, developers, and end users**.** Each of these.....
```

> Each of these roles has different privileges when connecting applications 

What is meant by **connecting applications**? This sounds a little vague.

sarathecoubian replied:
```suggestion
Connected Apps can be used by organization administrators, developers, and end users.
```

0xTenable commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,80 @@
+
+Organization administrators are the only users that can view and manage connected apps in Access Management.
+
+To manage your connected apps, navigate to Access Management and select the *Applications* tab.
```
Note: I think the **Applications** tab will be renamed to **Connected Apps** in the release

0xTenable commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,80 @@
+
+* An organization can own up to 200 connected apps.
+* You can disable or whitelist connected apps.
+* When you create an app, a clientID is assigned, which can't be modified. However, you can modify client secrets in the *Application Settings* page.
```
suggestion: a _unique_ clientId is assigned

sarathecoubian replied:
```suggestion
* When you create an app, a unique clientID is assigned; the clientID can't be modified. However, you can modify client secrets in the *Application Settings* page.
```

0xTenable commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,80 @@
+
+ The app directs the user's browser (commonly called a User-Agent) to Anypoint, where the user authorizes the app before being redirected back. If the code should not work indefinitely, you can select the *Refresh Tokens* option, requiring new codes as often as specified in the application.
+
+* JTW Bearer:
```
I think it's JWT Bearer

sarathecoubian replied:
```suggestion
* JWT Bearer:
```

sarathecoubian replied:
It definitely is - whoops. Good catch.

0xTenable reviewed (commented):
Initial comments

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+
+== Authentication
+
+Use OAuth 2.0 and OpenID to authorize apps to access resources using tokens or Single Sign-On (SSO).
```
OpenID -> OpenID Connect

sebastiankorol replied:
I don't understand the "tokens or SSO" part of this sentence. What are you trying to say?

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+== Authentication
+
+Use OAuth 2.0 and OpenID to authorize apps to access resources using tokens or Single Sign-On (SSO).
+Additionally, you can share account data with third-party applications.
```
Why is sharing data _additionally_ to accessing resources? your data is one of the resources that can be accessed

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+include::_attributes.adoc[]
+endif::[]
+
+You can connect apps to Anypoint Platform without granting them unrestricted access to data or having to share sensitive credentials.
```
OAuth is less about "connecting" apps, more about securely delegating user access to a client application with a restricted set of permissions called scopes. Can we make sure that's conveyed in this paragraph?

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+endif::[]
+
+You can connect apps to Anypoint Platform without granting them unrestricted access to data or having to share sensitive credentials.
+When you grant an app access, you have direct control over its data access, usage, and credentials.
```
You do not have control over credentials

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+
+You can connect apps to Anypoint Platform without granting them unrestricted access to data or having to share sensitive credentials.
+When you grant an app access, you have direct control over its data access, usage, and credentials.
+You can audit the use of data by connected apps, and you can also revoke access at any time.
```
You can audit actions taken by a connected app on your users' behalf

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+Additionally, you can share account data with third-party applications.
+For more information on which users can grant access to applications using Anypoint Platform, see Connected Apps Permissions by User.
+
+== First-party and Third-party Applications
```
nit: apps

sarathecoubian replied:
```suggestion
== First-party and Third-party Apps
```

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+
+== First-party and Third-party Applications
+
+To determine whether the app you want to connect is a first- or third-party application, consider the below:
```
This reads weird. There's multiple personas involved here -- There's an org admin who creates first-party clients, and whitelists third-party clients, and there's a user who authorizes a client application (i.e. connects the application). Users do not need to know or care if the app is first- or third-party.

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+
+  An app that is controlled by another organization.
+
+Applications can also be public or private.
```
Related to the above comment, categorizing first/third party next to public/private is weird because public/private is absolute, but parties are relative; what's a first-party app to me is a third-party app to you. That is, parties aren't a property of the app alone, while visibility is.

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,36 @@
+
+* Public app: 
+
+  An app that can be accessed by anyone.
```
Accessing an app is not a defined action. You either manage it, or you authorize it

sebastiankorol commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,80 @@
+include::_attributes.adoc[]
+endif::[]
+
+Connected Apps can be used by organization administrators, developers, and end users; each of these roles has different privileges when connecting applications,
```
please standardize on one term for connecting an application: Using / connecting / authorizing.

sebastiankorol commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,80 @@
+endif::[]
+
+Connected Apps can be used by organization administrators, developers, and end users; each of these roles has different privileges when connecting applications,
+and there are some restrictions to ensure sensitive data remains secure.
```
What restrictions? what sensitive data?

sebastiankorol commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,80 @@
+* An organization can own up to 200 connected apps.
+* You can disable or whitelist connected apps.
+* When you create an app, a clientID is assigned, which can't be modified. However, you can modify client secrets in the *Application Settings* page.
+* When you create an application with the OAuth client credential grant type, the app is owned by the organization; any organization administrator can control the app.
```
This is true of any grant type, not just client_credentials

sebastiankorol replied:
client_credentials

sebastiankorol commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,80 @@
+* You can disable or whitelist connected apps.
+* When you create an app, a clientID is assigned, which can't be modified. However, you can modify client secrets in the *Application Settings* page.
+* When you create an application with the OAuth client credential grant type, the app is owned by the organization; any organization administrator can control the app.
+* For apps that use the user grant type, actions taken by the app are logged with an application ID and user ID in the audit log so administrators can monitor and audit data use.
```
there's no "user grant type". There are several grant types that an application can use to act on behalf of a user: JWT Bearer, authorization_code, and basically every grant type that's not client_credentials

sebastiankorol commented on modules/ROOT/pages/connected-apps-permissions.adoc:
```diff
@@ -0,0 +1,80 @@
+
+* *Redirect URIs*:
+
+ Uniform Resource Identifier that specifies the location of the application's callback. You can specify one per line.
```
What's an application callback? Let's explain it further

sarathecoubian committed:
Apply suggestions from code review

sarathecoubian committed:
Restructure of connected apps

sebastiankorol commented on modules/ROOT/pages/connected-apps-devs-end-users.adoc:
```diff
@@ -0,0 +1,21 @@
+
+== Developers
+
+Developers who have websites or web apps can allow users to log in using the *Log in with Anypoint Platform* option.
```
Connected Apps does a lot more than Log In with Anypoint. It allows users of the platform to delegate authorization to third-party developer to act on their behalf within the platform

sarathecoubian replied:
```suggestion
Developers who have websites or web apps can allow users to log in using the *Log in with Anypoint Platform* option. Adding this client enables users of the website or web app to allow the third-party developer to act on their behalf.
```

sebastiankorol commented on modules/ROOT/pages/connected-apps-devs-end-users.adoc:
```diff
@@ -0,0 +1,21 @@
+== Developers
+
+Developers who have websites or web apps can allow users to log in using the *Log in with Anypoint Platform* option.
+They can also enable or disable the client, and they can view client usage.
```
What client? It's not defined perviously

Commit:
Update connected-apps-overview.adoc

@sarathecoubian made a few updates and changes. Could you please review/polish. 

@sebastiankorol feel free to drop your comments.

Commit:
Update connected-apps-overview.adoc

Commit:
Update connected-apps-devs-end-users.adoc

@sarathecoubian could you review? Also, let's add the screenshot - the UX is final in QAx.

@sebastiankorol please drop your comments.

Commit:
Update connected-apps-org-admin.adoc

@sarathecoubian could you review the formating? some sections show up with * instead of bold. 

@sebastiankorol feel free to drop comments.

sarathecoubian committed:
Small edits per style guide

sarathecoubian committed:
Corrections per style guide

sarathecoubian committed:
Corrections per style guide

dhirsch1 commented on modules/ROOT/pages/connected-apps-devs-end-users.adoc:
```diff
@@ -5,12 +5,10 @@ endif::[]
-For example, they can see how many organizations are using the app and how many users have access.
-
-// code/chunk of steps for websites/web apps
+In addition to enabling new flows to authenticate users, developers can enable or disable the clients and view client usage.
```
Suggest "clients" instead of "the clients"

sarathecoubian replied:
```suggestion
In addition to enabling new flows to authenticate users, developers can enable or disable clients and view client usage.
```

dhirsch1 commented on modules/ROOT/pages/connected-apps-devs-end-users.adoc:
```diff
@@ -19,4 +17,4 @@ They can can authorize access to their account when the following message appear
 // screenshot of authorize user page https://qax.anypoint.mulesoft.com/login/authorize?state=eyJjbGllbnQiOnsiY2xpZW50X2lkIjoiZWEzYzJiODFkMDEyNGI0Mjk1NDYyM2RhMWM0MzQ5ZWIifSwidXNlciI6eyJ1c2VybmFtZSI6Im9yZ2FuaXphdGlvbl9vd25lciJ9LCJvcmlnaW4iOiIvYXBpL3YyL29hdXRoMi9hdXRob3JpemU%2FY2xpZW50X2lkPWVhM2MyYjgxZDAxMjRiNDI5NTQ2MjNkYTFjNDM0OWViJnJlZGlyZWN0X3VyaT1odHRwczovL2V4YW1wbGUuY29tJnJlc3BvbnNlX3R5cGU9dG9r ...
 
-End users can manage third-party apps that they have authorized in their profile settings. To view autorized applications, click on the user's icon on the top right and navigate to profile. 
+End users can manage third-party apps that they have authorized in their profile settings. The user can view authorized applications in their profile.
```
Change "applications" to "apps." 
"The user" does not match "their" in the last sentence. Why did you remove the navigation information for viewing the apps?

sarathecoubian replied:
```suggestion
End users can manage third-party apps that they have authorized in their profile settings. The user can view authorized apps in their profile.
```

sarathecoubian replied:
Users know how to access their profiles, presumably, so I would prefer not to restate it.

dhirsch1 reviewed (changes requested):
A few small things

sarathecoubian committed:
Added Whitelisting

sarathecoubian commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,22 @@
+* A developer can register new and manage existing apps within their organization. 
+* An end user can authorize apps to access particular information, such as viewing assets in Exchange. 
+
+// will add links to the persona pages for each of those lines.
```
```suggestion

```

sarathecoubian committed:
Apply suggestions from code review

sarathecoubian committed:
Updated navigation

sarathecoubian committed:
Merge branch 'DOCS-5467-Access-Management-connected-apps' of https://github.com/mulesoft/docs-access-management into DOCS-5467-Access-Management-connected-apps

sebastiankorol commented on modules/ROOT/pages/connected-apps-devs-end-users.adoc:
```diff
@@ -0,0 +1,20 @@
+
+== Developers
+
+Third-party developers who have their own websites or web apps can allow users to log in using the *Log in with Anypoint Platform* option. For seamless implementation, there is a standard button that can be integrated into the front end of your application with a pre-built template and set of components. For more information about how to embed the button, refer to the  https://github.com/anypoint ...
```
This is simply one of the many use cases for oauth/oidc. Can first include a description of what Developers can do generically, before introducing Log in with Anypoint?

sebastiankorol replied:
nit: use https://github.com/anypoint-web-components/anypoint-signin rather than a link to the master branch

sebastiankorol replied:
Where did this surge.sh website come from? I'm concerned about endorsing it. It doesn't even support https

sebastiankorol commented on modules/ROOT/pages/connected-apps-devs-end-users.adoc:
```diff
@@ -0,0 +1,20 @@
+
+== End Users
+
+End users can log into a third-party applications using their Anypoint credentials.
```
Same as above, this enables more use cases other than log in with anypoint

sebastiankorol commented on modules/ROOT/pages/connected-apps-devs-end-users.adoc:
```diff
@@ -0,0 +1,20 @@
+End users can log into a third-party applications using their Anypoint credentials.
+They can can authorize access to their account when the following message appears:
+
+// screenshot of authorize user page https://qax.anypoint.mulesoft.com/login/authorize?state=eyJjbGllbnQiOnsiY2xpZW50X2lkIjoiZWEzYzJiODFkMDEyNGI0Mjk1NDYyM2RhMWM0MzQ5ZWIifSwidXNlciI6eyJ1c2VybmFtZSI6Im9yZ2FuaXphdGlvbl9vd25lciJ9LCJvcmlnaW4iOiIvYXBpL3YyL29hdXRoMi9hdXRob3JpemU%2FY2xpZW50X2lkPWVhM2MyYjgxZDAxMjRiNDI5NTQ2MjNkYTFjNDM0OWViJnJlZGlyZWN0X3VyaT1odHRwczovL2V4YW1wbGUuY29tJnJlc3BvbnNlX3R5cGU9dG9r ...
```
Replace with screenshot

sebastiankorol commented on modules/ROOT/pages/connected-apps-devs-end-users.adoc:
```diff
@@ -0,0 +1,20 @@
+
+// screenshot of authorize user page https://qax.anypoint.mulesoft.com/login/authorize?state=eyJjbGllbnQiOnsiY2xpZW50X2lkIjoiZWEzYzJiODFkMDEyNGI0Mjk1NDYyM2RhMWM0MzQ5ZWIifSwidXNlciI6eyJ1c2VybmFtZSI6Im9yZ2FuaXphdGlvbl9vd25lciJ9LCJvcmlnaW4iOiIvYXBpL3YyL29hdXRoMi9hdXRob3JpemU%2FY2xpZW50X2lkPWVhM2MyYjgxZDAxMjRiNDI5NTQ2MjNkYTFjNDM0OWViJnJlZGlyZWN0X3VyaT1odHRwczovL2V4YW1wbGUuY29tJnJlc3BvbnNlX3R5cGU9dG9r ...
+
+End users can manage third-party apps that they have authorized in their profile settings. The user can view authorized apps in their profile.
```
nit: First sentence is plural, second is singular

sebastiankorol replied:
Can you show a screenshot and instructions on how to get to the profile settings?

sebastiankorol commented on modules/ROOT/pages/connected-apps-devs-end-users.adoc:
```diff
@@ -0,0 +1,20 @@
+include::_attributes.adoc[]
+endif::[]
+
+== Developers
```
Please link and explain https://anypoint.mulesoft.com/accounts/api/v2/oauth2/.well-known/openid-configuration -- Developers needs this as well as the URLs that it contains to function properly.

An e2e example of how to set up a connected app as a Developer would be useful here too.

Please also list what parts of the spec and protocols we're compliant with

sebastiankorol replied:
Please add an explicit mention of the refresh_token scope, what it is, and how to use it to obtain new access tokens

sebastiankorol replied:
As written, this assumes the user has complete understanding of OIDC, as there's no mention of access tokens, ID Tokens, or any other critical piece of the protocol.

sebastiankorol replied:
Linking https://anypoint.mulesoft.com/exchange/portals/anypoint-platform/ here seems critical, since those are the APIs developers will interact with

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,21 @@
+include::_attributes.adoc[]
+endif::[]
+
+Connected Apps provide a framework that enables an external application to integrate with MuleSoft using APIs via standard protocols, such as SAML, OAuth, and OpenID Connect. Connected apps help user delegate their access without sharing sensitive credentials or giving full control of their account to third parties. Actions taken by connected apps are auditable, and you can also revoke access at  ...
```
Connected Apps are an OAuth 2.0 / OIDC implementation. SAML isn't supported by Connected Apps

sebastiankorol replied:
nit: Connected apps help user -> Connected apps help users. Also the paragraph switches from the third person to the second person

sebastiankorol commented on modules/ROOT/pages/connected-apps-org-admin.adoc:
```diff
@@ -0,0 +1,76 @@
+To manage your connected apps, navigate to Access Management and select the *Connected Apps* tab.
+You can add apps in the *Owned Apps* section and manage app access in the *Authorized Apps* section.
+
+=== Create an App
```
Shouldn't create an app be in the Developer persona section of the docs?

sebastiankorol commented on modules/ROOT/pages/connected-apps-overview.adoc:
```diff
@@ -0,0 +1,21 @@
+
+Connected Apps provide a framework that enables an external application to integrate with MuleSoft using APIs via standard protocols, such as SAML, OAuth, and OpenID Connect. Connected apps help user delegate their access without sharing sensitive credentials or giving full control of their account to third parties. Actions taken by connected apps are auditable, and you can also revoke access at  ...
+
+== Authentication
```
I'm a little bit confused by this section, as the information here is terse and contained in the other sections mostly

sebastiankorol reviewed (changes requested):
Let's flesh out the Developer section please

sarathecoubian committed:
DOCS-5467-revisions

sarathecoubian committed:
nav

sarathecoubian committed:
Update connected-apps-end-users.adoc

sarathecoubian committed:
images modified
